engine/common/splitter/network/network.go

```diff
@@ -0,0 +1,96 @@
+package network
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	splitterEngine "github.com/onflow/flow-go/engine/common/splitter"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/network"
+	"github.com/rs/zerolog"
+)
+
+type Network struct {
+	mu        sync.RWMutex
+	net       module.Network
+	log       zerolog.Logger
+	splitters map[network.Channel]*splitterEngine.Engine // stores splitters for each channel
+	conduits  map[network.Channel]network.Conduit        // stores conduits for all registered channels
+	subMngr   network.SubscriptionManager
+}
+
+func NewNetwork(
+	net module.Network,
+	log zerolog.Logger,
+	subMngr network.SubscriptionManager,
+) (*Network, error) {
+	e := &Network{
+		net:       net,
+		splitters: make(map[network.Channel]*splitterEngine.Engine),
+		conduits:  make(map[network.Channel]network.Conduit),
+		log:       log,
+		subMngr:   subMngr,
+	}
+
+	return e, nil
+}
+
+// Register will subscribe the given engine with the spitter on the given channel, and all registered
+// engines will be notified with incoming messages on the channel.
+// The returned Conduit can be used to send messages to engines on other nodes subscribed to the same channel
+func (n *Network) Register(channel network.Channel, e network.Engine) (network.Conduit, error) {
+	engine, ok := e.(module.Engine)
+
+	if !ok {
```

Comment:
may be also add the check,
`engine, ok := e.(splitter.Engine)`
(one should not attempt to register a splitter engine with the splitter network)

Reply:
Perhaps, but I don't think it's necessary actually. 

I can think of cases where one may actually want to register a splitter engine with the splitter network.

For example, one cannot access the splitter engines that are managed by the splitter network. They are completely encapsulated within the splitter network instance, who is the only one that can call the `RegisterEngine` and `UnregisterEngine` methods on them. Therefore, one may wish to register their own splitter engine with the splitter network (or any network in general) since it would allow them to dynamically register or unregister engines with the splitter.

Right now there probably aren't such usecases, but overall I wouldn't consider it an error if someone tried to register a splitter engine with the splitter network :)